File: techila/examples/R/Advanced/Asian/run_asian_cloudfor.r
# Copyright 2010-2013 Techila Technologies Ltd.

# This R-script contains the Local Control Code, which will be used to
# distribute computations to the Techila environment using cloudfor.
#
# The R-script named "asian_function.r" will be sourced and call to
# the "asian_function" is used inside the cloudfor loops.
#
# Results will be visualized by displaying a graph on the screen.
#
# Usage:
#
# source("run_asian_cloudfor.r")
# price <- run_asian_cloudfor()

library(techila)

total <- new.env()

run_asian_cloudfor<- function() {
  source("parameters.r")

  source("plotresults.r")
  source("asian_function.r")

  cloudfor(i = 1:length(S0)) %to%
  cloudfor(j = 1:length(sigma0),
           .callback = plotresults,
           .stream = TRUE,
           .sdkroot="../../../../"
           ) %t% {

    jobprice <- asian_function(S0[i], sigma0[j]^2, M, nn, r, N, rho, kappa, psi, E, T)
    list(i_index=i, j_index=j, price=jobprice)
  }

  total$price
}

File: techila/examples/Matlab/Advanced/Asian/run_asian_cloudfor.m
function run_asian_cloudfor()
% This function shows how you can distribute the computationally intensive
% Asian routine in the 'asian_function' to the Workers by using the
% cloudfor-function.
%
%Usage: run_asian_cloudfor()
S0 = linspace(45,47,9); % initial stock price
sigma0 = linspace(0.35,0.4,9);%initial volatility of stock return

%The parameters of stock price diffusion
M = 20000; %The number of trajectories
N = 365; %The number of data points in year
nn=1;    % number of time steps per sample point

%The parameters of volatility diffusion
rho=-0.5; %The correlation between the increments of stock price and volatility
kappa=0.1; %The speed of revision
psi=0.5; %The standard deviation of volatility

%Other parameters
E=70; %Exercise price
T=1; %Maturity time
r = 0.05; % Interest rate

price = zeros(length(sigma0),length(S0));

[l_grid S2_grid] = meshgrid(S0,sigma0);
%price = nan(size(l_grid));
price = nan(length(sigma0),length(S0));
figure;
fig = surf(l_grid,S2_grid,price,'ZDataSource','price');
colormap(jet)

xlabel('Initial stock price'),ylabel('Initial volatility'),zlabel('Price of Asian Call')
title('Plot generated by the distributed version (cloudfor)')
drawnow;

cloudfor i_index=1:length(S0)
%cloudfor('stepsperjob',1)
        cloudfor j_index=1:length(sigma0)
        %cloudfor('stepsperjob',1)
            price(j_index,i_index) = asian_function(S0(i_index),sigma0(j_index)^2,M,nn,r,N,rho,kappa,0.5,E,T);
            %cloudfor('callback','refreshdata(fig, 'caller');')
            %cloudfor('callback','drawnow;')
        cloudend
cloudend

refreshdata(fig, 'caller');
drawnow;
